db/memtable_list.cc

```diff
@@ -240,19 +245,43 @@ void MemTableListVersion::Remove(MemTable* m,
   memlist_.remove(m);
 
   m->MarkFlushed();
-  if (max_write_buffer_number_to_maintain_ > 0) {
+  if (max_write_buffer_size_to_maintain_ > 0 ||
+      max_write_buffer_number_to_maintain_ > 0) {
     memlist_history_.push_front(m);
     TrimHistory(to_delete);
   } else {
     UnrefMemTable(to_delete, m);
   }
 }
 
+size_t MemTableListVersion::ApproximateMemoryUsage() {
+  // use *parent_memtable_list_memory_usage_ instead?
+  size_t total_memtable_size = 0;
+  for (auto& memtable : memlist_) {
+    total_memtable_size += memtable->ApproximateMemoryUsage();
+  }
+  for (auto& memtable : memlist_history_) {
+    total_memtable_size += memtable->ApproximateMemoryUsage();
+  }
+  return total_memtable_size;
+}
+
+bool MemTableListVersion::MemtableLimitExceeded(size_t usage) {
+  if (max_write_buffer_size_to_maintain_ > 0) {
+    return ApproximateMemoryUsage() + usage >
+           static_cast<size_t>(max_write_buffer_size_to_maintain_);
```

Comment:
How would the user be expected to configure size to maintain in the example given by @siying, i.e they want to maintain a history of 64MB, but the max memtable size is 256MB? IIUC, with this implementation, they would set it to 320MB. IMO, it would be more intuitive for them to set it to 64MB and we treat it as a soft limit, i.e keep accumulating memtables until there is a minimum of 64MB

Reply:
With commit 1222861 we now exclude the last MemTable in memlist_history_ from memory usage calculation, so for the case of 256+64, `ApproximateMemoryUsageExcludingLast` now returns 0 and `usage` is 64, so here 64 == 64, we will drop the flushed memtable and reduce the total memory usage back to 64. One thing that's open to debate is whether we should make `max_write_buffer_size_to_maintain` an open or closed upper bound. With the current approach, immutable memtable size will never reach `max_write_buffer_size_to_maintain`, but some user may assume it's an closed range